components/ee/ws-scheduler/pkg/scaler/driver.go

```diff
@@ -0,0 +1,380 @@
+		Controller: controller,
+		stop:       make(chan struct{}),
+	}, nil
+}
+
+// WorkspaceManagerPrescaleDriver implements a prescale driver using ws-manager's ghost pods
+type WorkspaceManagerPrescaleDriver struct {
+	Config WorkspaceManagerPrescaleDriverConfig
+
+	Client api.WorkspaceManagerClient
+	conn   *grpc.ClientConn
+
+	Controller Controller
+
+	stop chan struct{}
+	once sync.Once
+}
+
+type workspaceStatus struct {
+	Count              WorkspaceCount
+	DeletionCandidates []string
+}
+
+// Run runs the prescale driver until Stop() is called
+func (wspd *WorkspaceManagerPrescaleDriver) Run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	statusChan := make(chan workspaceStatus)
+	go func() {
+		for {
+			err := wspd.maintainWorkspaceStatus(ctx, statusChan)
+			if err == context.Canceled {
+				return
+			}
+			if err != nil {
+				log.WithError(err).Error("cannot maintain workspace count")
+				time.Sleep(1 * time.Second)
+			}
+		}
+	}()
+
+	var renewal <-chan time.Time
+	if wspd.Config.Renewal.Percentage > 0 && wspd.Config.Renewal.Interval > 0 {
+		t := time.NewTicker(time.Duration(wspd.Config.Renewal.Interval))
+		defer t.Stop()
+		renewal = t.C
+	}
+	houseKeeping := time.NewTicker(1 * time.Minute)
+	defer houseKeeping.Stop()
+
+	var (
+		counts         = make(chan WorkspaceCount)
+		status         workspaceStatus
+		startingGhosts = make(map[string]time.Time)
+	)
+
+	delta := wspd.Controller.Control(ctx, counts)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case status = <-statusChan:
+			for _, id := range status.DeletionCandidates {
+				delete(startingGhosts, id)
+			}
+			status.Count.Ghost += len(startingGhosts)
+			counts <- status.Count
+			log.WithField("counts", counts).Debug("status update")
+		case <-houseKeeping.C:
+			for id, t1 := range startingGhosts {
+				if time.Since(t1) < maxGhostStartTime {
+					continue
+				}
+				delete(startingGhosts, id)
+				status.Count.Ghost--
+			}
+		case <-renewal:
+			d := int(float64(len(status.DeletionCandidates)) * float64(wspd.Config.Renewal.Percentage) * 0.01)
+			if d == 0 {
+				continue
+			}
+			if d > len(status.DeletionCandidates) {
+				d = len(status.DeletionCandidates)
+			}
+			log.WithField("count", d).Info("attempting to renew ghost workspaces")
+			err := wspd.stopGhostWorkspaces(ctx, status.DeletionCandidates[:d])
+			if err != nil {
+				log.WithError(err).Error("cannot stop ghost workspaces during renewal")
+				continue
+			}
+			ids, err := wspd.startGhostWorkspaces(ctx, d, status)
+			if err != nil {
+				log.WithError(err).Error("cannot start ghost workspaces during renewal")
+				continue
+			}
+			for _, id := range ids {
+				startingGhosts[id] = time.Now()
+				status.Count.Ghost++
+			}
+		case d := <-delta:
+			if d == 0 {
+				continue
+			}
+			if wspd.Config.ReactionDelay > 0 {
+				time.Sleep(time.Duration(wspd.Config.ReactionDelay))
+			}
+			var (
+				err error
+				ids []string
+			)
```

Comment:
Uff. The following passage has multiple intended "fallthroughs" which make it more compact but not straight forward to read... feels a tad to "clever" for my taste.

Reply:
Unfortunately I don't see this. I've added a couple of breaklines to try and make the structure clearer.
Did that help?

Reply:
Yes, better.

Just to explain what I meant, no need to take action:
 - `if ... if` in comparison to `if ... else`: The latter better guides the eye IMO (taste)
 - both cases can definitely result in errors. Having the fallthrough is nice because it reduces dublication but also needs another "eye-jump" to parse correctly - especially if combined with the `if ... if` from above (again, taste)